### DISCUSSION TITLE
Add "topupInvoice" to checkout parameters for an invoice

### DISCUSSION THREAD

**Discussion by ashleymcdonald** [btcpayserver/btcpayserver#5759] [Ideas / Feature Requests]
Feature request to add "topupInvoice" (bool) with default false to checkout parameters for an invoice

This would enable the mode where any payment is considered final, allowing top-up invoices to have a suggested amount.

The flow would be as follows;

- topup invoice created for $100 => 0.000019BTC
  - user sends the correct amount and invoice works as normal
  - OR user sends a different amount
    - if its within the paymentTolerance then invoice acts as normal
    - if its not within paymentTolerance then **invoice value is changed** to converted amount received and then **continues as normal** (security settings etc determine the flow from here)

This addresses cases where top-ups should have a suggested value to help users who do not need/want to conversions and ensure the user gets the amount they requested whilst also allowing users to send any other amount to fulfil the request.

*Note user may also send more than requested

**Comment by Kukks** [comment]
I don't quite understand. The feature you want is called a topup invoice, where you allow 1 payment to be sent, which then sets the invoice amount based on the exchange rate set at invoice creation.

What you're asking for is maybe offering a hint to the user that if they send X amount of BTC, it is equivalent to Y USD?
![image](https://github.com/btcpayserver/btcpayserver/assets/1818366/2f2398ba-07fe-4118-96af-4574f1a4b21c)
I mean you can do a set amount or you can do "any" amount. But not both. Perhaps you can show both options prior to creating the invoice similar to what opensats does.
![](https://i.imgur.com/07JIbI2.gif)

**Reply by ashleymcdonald** [reply]
ok in that example, what happens when the user chooses to donate 100 and sends 50?

for a donation its not as important as credit to an account though.

**Reply by ashleymcdonald** [reply]
Example flow:

In the account interface in the store, the user wants to add credit to his account; 
- he chooses 200 and clicks next. 
- he then looks in his wallet and sees he doesn't have 200
- but since it's store credit, he sends 100 anyway. 

The change I'm proposing is that when the invoice is set as a "top-up" then it will accept the 100 as the total amount, change the value of the invoice to 100 and continue as normal, keeping any security aspects in place like confirmations etc. and having the invoices match in both systems and so on.

**Reply by ashleymcdonald** [reply]
Also, it's important to note that the majority send the correct amount, and having the BTC amount on the payment page is important to the majority.

The users who decide to send a different amount due to whatever reason break the "set amount" methodology as a boolean; this would solve the issue and also enable other options in the future for places that use the server as a way to func credit based accounts or tokens of X denomination. 

This could possibly, instead of being a "Topup" flag, be a "Accept any amount" flag. 
It does need to update the invoice total to the accepted amount for it to play well with integrated systems

**Reply by Kukks** [reply]
Ok, I get it. 

We actually do have a property, `invoiceType`, which explicitly sets the invoice as standard (with amount) or Topup (amountless).

so the proposal here is to ensure you can set an initial target amount but also be ok if the amount is set to something else.

Might require a decent amount of work to ensure no bugs and corner cases. Also, since you mentioned paymentTolerance, using this with "account topup" flows might bring up some gaming exploitation by users, who will just consistently send the absolute minimum to hit the tolerance and save the %.

**Reply by ashleymcdonald** [reply]
I detailed the tolerance concept here #5761 this is more for a multiple on topups. With tolerance being a store setting; leaving it at 0 removes the store risk in exchange for not pretty numbers :P

Not that I subscribe to security through obscurity, but your example would mean the user needs to know the tolerance settings. is this public info?

**Reply by Kukks** [reply]
It's not but you can just iterate until you find the sweet spot

**Reply by ashleymcdonald** [reply]
There are users exploiting the payment tolerance already, this change would allow it to be set to 0 and just credit the real amounts, thus educating the users that they should send the total amount to get the quoted value. 
We only use the payment tolerance because it costs the business less than the manual review and CS costs.

**Reply by ashleymcdonald** [reply]
> It's not but you can just iterate until you find the sweet spot

hah if users want to keep sending money to find how to save some then most business models would call that a loyalty discount :P
